scipy/optimize/_differentialevolution.py

```diff
@@ -598,6 +619,43 @@ def __init__(self, func, bounds, args=(),
 
         self.random_number_generator = check_random_state(seed)
 
+        # Which parameters are going to be integers?
+        if np.any(integrality):
+            # user has provided a truth value for integer constraints
+            if polish:
+                message = ("Polishing is incompatible with integrality"
+                           " constraints; `polish` will be ignored.")
+                self.polish = False
+                warnings.warn(message, OptimizeWarning)
+
+            self.integrality = np.broadcast_to(
+                integrality,
+                self.parameter_count
+            )
+            self.integrality = np.asarray(self.integrality, bool)
+            # For integrality parameters change the limits to only allow
+            # integer values lying between the limits.
+            lb, ub = np.copy(self.limits)
+
+            lb[lb > 0] = np.ceil(lb[lb > 0])
+            lb[lb < 0] = np.trunc(lb[lb < 0])
+            ub[ub > 0] = np.trunc(ub[ub > 0])
+            ub[ub < 0] = np.floor(ub[ub < 0])
+            if not (lb[integrality] <= ub[integrality]).all():
+                # there's a parameter that doesn't have an integer value
+                # lying between the limits
+                raise ValueError("One of the integrality constraints does not"
+                                 " have any possible integer values between"
+                                 " the lower/upper bounds.")
+            self.limits[0, self.integrality] = (
+                    lb[self.integrality] - (0.5 - _MACHEPS)
+            )
+            self.limits[1, self.integrality] = (
+                    ub[self.integrality] + (0.5 - _MACHEPS)
+            )
```

Comment:
Why did you choose between `ceil`/`trunc` or `floor`/`trunc` based on the sign? I think `ceil` and `floor` work as we want them to regardless of the sign.

Also, I think you only need to "modify" the bounds of the integrality constraints ,right? Conceptually (not being perfectly precise, and not being careful about machine precision issues) this is all:
```suggestion
            lb[self.integrality] = np.ceil(lb[self.integrality]) - 0.4999
            ub[self.integrality] = np.floor(ub[self.integrality]) + 0.4999
            if np.any(np.round(ub[self.integrality]) - np.round(lb[self.integrality]) < 1):
                raise ValueError(message)
```
If we're careful, though, rather than adding or subtracting _MACHEPS, it would be safer to do multiplication/division by a number very close to 1, I think. For instance, 
```
>>> 1e300 + _MACHEPS > 1e300
False
>>> 1e300 - _MACHEPS < 1e300
False
>>> 1e300*(1 + _MACHEPS) > 1e300
True
>>> 1e300*(1 - _MACHEPS) < 1e300
True
```
Of course, if the magnitudes we're dealing with are larger than the largest integer that can represented exactly in `float64`, it's not going to work perfectly either way.

Practically speaking, I wouldn't object to actually using the code above. It's easier to understand, and it will only reduce the chance that the integers on the bounds are selected by a very small amount.

Reply:
I'll use `np.nextafter`, that will do the job.

Reply:
Wow, what a handy function. Didn't know about that one. 
I'm not sure it's needed because the RNG is [0, 1) but I guess it doesn't hurt.